experimental/ir/fdp.go

```diff
@@ -0,0 +1,443 @@
+		fdp.Type = kind.Enum()
+	} else if ty.IsEnum() {
+		fdp.Type = descriptorpb.FieldDescriptorProto_TYPE_ENUM.Enum()
+		fdp.TypeName = addr(ty.Name())
+	} else {
+		// TODO: Groups
+		fdp.Type = descriptorpb.FieldDescriptorProto_TYPE_MESSAGE.Enum()
+		fdp.TypeName = addr(ty.Name())
+	}
+
+	if f.IsExtension() {
+		fdp.Extendee = addr(f.Container().Name())
+	}
+
+	if oneof := f.Oneof(); !oneof.IsZero() {
+		fdp.OneofIndex = addr(int32(oneof.Index()))
+	}
+
+	for option := range seq.Values(f.Options()) {
+		if fdp.Options == nil {
+			fdp.Options = new(descriptorpb.FieldOptions)
+		}
+
+		// We pass the FDP directly, because we want to keep it around for
+		// dealing with the pseudo-options. codegen.option has a special case
+		// for this.
+		c.option(option, fdp)
+	}
+}
+
+func (c *codegen) oneof(o Oneof, odp *descriptorpb.OneofDescriptorProto) {
+	odp.Name = addr(o.Name())
+
+	for option := range seq.Values(o.Options()) {
+		if odp.Options == nil {
+			odp.Options = new(descriptorpb.OneofOptions)
+		}
+
+		c.option(option, odp.Options)
+	}
+}
+
+func (c *codegen) enum(ty Type, mdp *descriptorpb.EnumDescriptorProto) {
+	mdp.Name = addr(ty.Name()) // Has a leading dot.
+
+	for field := range seq.Values(ty.Fields()) {
+		evd := new(descriptorpb.EnumValueDescriptorProto)
+		mdp.Value = append(mdp.Value, evd)
+		c.enumValue(field, evd)
+	}
+
+	//nolint:revive // Complains about range_.
+	for range_ := range seq.Values(ty.ReservedRanges()) {
+		rr := new(descriptorpb.EnumDescriptorProto_EnumReservedRange)
+		mdp.ReservedRange = append(mdp.ReservedRange, rr)
+
+		start, end := range_.Range()
+		rr.Start = addr(start)
+		rr.End = addr(end - 1)
+	}
+
+	for name := range seq.Values(ty.ReservedNames()) {
+		mdp.ReservedName = append(mdp.ReservedName, name.Name())
+	}
+
+	for option := range seq.Values(ty.Options()) {
+		if mdp.Options == nil {
+			mdp.Options = new(descriptorpb.EnumOptions)
+		}
+
+		c.option(option, mdp.Options)
+	}
+}
+
+func (c *codegen) enumValue(f Field, fdp *descriptorpb.EnumValueDescriptorProto) {
+	fdp.Name = addr(f.Name())
+	fdp.Number = addr(f.Number())
+
+	for option := range seq.Values(f.Options()) {
+		if fdp.Options == nil {
+			fdp.Options = new(descriptorpb.EnumValueOptions)
+		}
+
+		c.option(option, fdp.Options)
+	}
+}
+
+func (c *codegen) option(_ Option, target proto.Message) {
+	var fdp *descriptorpb.FieldDescriptorProto
+	if actual, ok := target.(*descriptorpb.FieldDescriptorProto); ok {
+		fdp = actual
+		target = fdp.Options
+	}
+
+	_ = target
+
+	// There are two cases and both are painful.
+	//
+	// 1. For built-in options, we need to match up option.Field() to a
+	//    protoreflect.Field in target, and then set it.
+	//
+	//    If we recognize this field as a pseudo-option, we need to forgo the
+	//    above and set it directly on the non-nil fdp instead.
+	//
+	// 2. For custom options, we need to serialize option (perhaps with an
+	//    Option.Marshal() function?) and append it to the unknown fields.
```

Comment:
We actually need to build a model of the value, like a dynamic message, in order to detect fields or oneofs that are set more than once (which is an error). We could have that model serialize itself to bytes at the end or we could just serialize/append them as we go (which I suppose means the model can be simpler since it's just book-keeping instead of actually accumulating all of the values/data).

Reply:
Is the idea that they will already be typed-checked? We probably also need a way to emit unrecognized options into the `unrecognized` fields of the various option types (which is admittedly much less painful... almost trivial in fact).

Reply:
Correct. There is a followup PR that performs the lowering out of Value, on the presumption type-checking has already happened and succeeded.